File: apps/frontend/src/lib/api-client/pinnedMessages.ts
import { RoomService } from '@chatto/api-types/api/v1/rooms_connect';
import type { PinnedMessage } from '@chatto/api-types/api/v1/rooms_pb';
import { authHeaders, createChattoClient, handleAuthError, type ConnectAPIConfig } from './connect';
import { timelineUsersForMessages } from './roomTimeline';

export type PinnedMessagesPage = {
  items: PinnedMessage[];
  totalCount: number;
  hasMore: boolean;
  latestPinMarker: string;
};

export function createPinnedMessagesAPI(config: ConnectAPIConfig) {
  const rooms = createChattoClient(RoomService, config);
  const headers = () => authHeaders(config);
  return {
    async list(roomId: string, limit: number, offset: number): Promise<PinnedMessagesPage> {
      try {
        const response = await rooms.listPinnedMessages(
          { roomId, page: { limit, offset } },
          { headers: headers() }
        );
        await timelineUsersForMessages(
          config,
          response.pinnedMessages.flatMap((item) => (item.message ? [item.message] : []))
        );
        return {
          items: response.pinnedMessages,
          totalCount: Number(response.page?.totalCount ?? response.pinnedMessages.length),
          hasMore: response.page?.hasMore ?? false,
          latestPinMarker: response.latestPinMarker
        };
      } catch (error) {
        return handleAuthError(config, error);
      }
    },
    async create(roomId: string, messageEventId: string): Promise<PinnedMessage | null> {
      try {
        const response = await rooms.createPinnedMessage(
          { roomId, messageEventId },
          { headers: headers() }
        );
        return response.pinnedMessage ?? null;
      } catch (error) {
        return handleAuthError(config, error);
      }
    },
    async remove(roomId: string, messageEventId: string): Promise<void> {
      try {
        await rooms.deletePinnedMessage({ roomId, messageEventId }, { headers: headers() });
      } catch (error) {
        return handleAuthError(config, error);
      }
    }
  };
}

export type PinnedMessagesAPI = ReturnType<typeof createPinnedMessagesAPI>;

File: cli/internal/connectapi/pinned_messages.go
package connectapi

import (
	"context"
	"errors"

	"connectrpc.com/connect"
	"hmans.de/chatto/internal/core"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
)

const (
	defaultPinnedMessageListLimit = 50
	maxPinnedMessageListLimit     = 100
)

func (s *roomService) ListPinnedMessages(ctx context.Context, req *connect.Request[apiv1.ListPinnedMessagesRequest]) (*connect.Response[apiv1.ListPinnedMessagesResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	limit, offset := apiPagination(req.Msg.GetPage(), defaultPinnedMessageListLimit, maxPinnedMessageListLimit)
	result, err := s.api.core.RoomTimelineReads().ListPinnedMessages(ctx, core.PinnedMessageListInput{ActorID: caller.UserID, RoomID: req.Msg.GetRoomId(), Limit: limit, Offset: offset})
	if err != nil {
		return nil, connectError(err)
	}
	items, err := newPinnedMessageAssembler(s.api).assemble(ctx, caller.UserID, result.Items)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.ListPinnedMessagesResponse{
		PinnedMessages:  items,
		Page:            apiPageInfo(result.TotalCount, result.HasMore),
		LatestPinMarker: result.LatestPinEventID,
	}), nil
}

func (s *roomService) CreatePinnedMessage(ctx context.Context, req *connect.Request[apiv1.CreatePinnedMessageRequest]) (*connect.Response[apiv1.CreatePinnedMessageResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	pin, err := s.api.core.RoomCommands().CreatePinnedMessage(ctx, core.PinnedMessageMutationInput{ActorID: caller.UserID, RoomID: req.Msg.GetRoomId(), MessageEventID: req.Msg.GetMessageEventId()})
	if err != nil {
		return nil, connectError(err)
	}
	result, err := s.api.core.RoomTimelineReads().GetMessage(ctx, caller.UserID, req.Msg.GetRoomId(), pin.MessageEventID)
	if err != nil {
		return nil, connectError(err)
	}
	pinned, err := newPinnedMessageAssembler(s.api).assemble(ctx, caller.UserID, []core.PinnedMessageItem{{Pin: pin, Event: result.Event}})
	if err != nil {
		return nil, connectError(err)
	}
	if len(pinned) != 1 {
		return nil, connectError(errors.New("pinned message hydration returned no result"))
	}
	return connect.NewResponse(&apiv1.CreatePinnedMessageResponse{PinnedMessage: pinned[0]}), nil
}

func (s *roomService) DeletePinnedMessage(ctx context.Context, req *connect.Request[apiv1.DeletePinnedMessageRequest]) (*connect.Response[apiv1.DeletePinnedMessageResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	deleted, err := s.api.core.RoomCommands().DeletePinnedMessage(ctx, core.PinnedMessageMutationInput{ActorID: caller.UserID, RoomID: req.Msg.GetRoomId(), MessageEventID: req.Msg.GetMessageEventId()})
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.DeletePinnedMessageResponse{Deleted: deleted}), nil
}
